File: implementations/java/provekit-lsp-java/src/test/java/com/provekit/lsp/ForwardPropagatorTests.java
package com.provekit.lsp;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class ForwardPropagatorTests {
    @Test
    void callsiteSatisfiesPre_noDiagnostic() {
        ForwardPropagator fp = new ForwardPropagator();
        fp.addToCatalog("checkPositive", 
            new ForwardPropagator.Post(java.util.List.of("x > 0"), false),
            new ForwardPropagator.Post(java.util.List.of("x <= 0"), false));
        
        var currentPost = new ForwardPropagator.Post(java.util.List.of("x > 0"), false);
        var result = fp.checkCallsite("checkPositive", currentPost);
        
        assertNull(result);
    }
    
    @Test
    void callsiteViolatesPre_diagnosticEmitted() {
        ForwardPropagator fp = new ForwardPropagator();
        fp.addToCatalog("checkPositive", 
            new ForwardPropagator.Post(java.util.List.of("x > 0"), false),
            new ForwardPropagator.Post(java.util.List.of("x <= 0"), false));
        
        var currentPost = new ForwardPropagator.Post(java.util.List.of("x <= 0"), false);
        var result = fp.checkCallsite("checkPositive", currentPost);
        
        assertNotNull(result);
        assertEquals("implication-failed", result.code);
    }
    
    @Test
    void topFallback_noDiagnostic() {
        ForwardPropagator fp = new ForwardPropagator();
        
        var currentPost = new ForwardPropagator.Post(java.util.List.of(), true);
        var result = fp.checkCallsite("checkPositive", currentPost);
        
        assertNull(result);
    }
}

File: implementations/swift/Sources/Provekit/ForwardPropagatorTests.swift
import Foundation
import XCTest
@testable import Provekit

final class ForwardPropagatorTests: XCTestCase {
    func test_callsiteSatisfiesPre_noDiagnostic() {
        let fp = ForwardPropagator()
        fp.addToCatalog("checkPositive", pre: .of("x > 0"), post: .of("x <= 0"))
        
        let currentPost = ForwardPropagator.Post.constraints(["x > 0"], isTop: false)
        let result = fp.checkCallsite("checkPositive", currentPost: currentPost)
        
        XCTAssertNil(result)
    }
    
    func test_callsiteViolatesPre_diagnosticEmitted() {
        let fp = ForwardPropagator()
        fp.addToCatalog("checkPositive", pre: .of("x > 0"), post: .of("x <= 0"))
        
        let currentPost = ForwardPropagator.Post.constraints(["x <= 0"], isTop: false)
        let result = fp.checkCallsite("checkPositive", currentPost: currentPost)
        
        XCTAssertNotNil(result)
        XCTAssertEqual(result?.code, "implication-failed")
    }
    
    func test_topFallback_noDiagnostic() {
        let fp = ForwardPropagator()
        fp.addToCatalog("checkPositive", pre: .of("x > 0"), post: .of("x <= 0"))
        
        let currentPost = ForwardPropagator.Post.constraints([], isTop: true)
        let result = fp.checkCallsite("checkPositive", currentPost: currentPost)
        
        XCTAssertNil(result)
    }
    
    func test_unknownCallee_noDiagnostic() {
        let fp = ForwardPropagator()
        
        let currentPost = ForwardPropagator.Post.constraints(["x > 0"], isTop: false)
        let result = fp.checkCallsite("unknown", currentPost: currentPost)
        
        XCTAssertNil(result)
    }
}